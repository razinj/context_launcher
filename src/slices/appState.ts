import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { AppDetails } from '../models/app-details'
import { RenderedIn } from '../models/rendered-in'
import { RootState } from '../store'

interface AppsSearchState {
  query?: string
  result: AppDetails[]
}

export interface AppState {
  displayAllApps: boolean
  displaySettings: boolean
  menuAppDetails?: AppDetails
  displayAppMenu: boolean
  displaySortableFavoriteApps: boolean
  displaySortablePinnedApps: boolean
  displaySortableTemporaryPinnedApps: boolean
  search: AppsSearchState
}

const initialState: AppState = {
  displayAllApps: false,
  displaySettings: false,
  menuAppDetails: undefined,
  displayAppMenu: false,
  displaySortableFavoriteApps: false,
  displaySortablePinnedApps: false,
  displaySortableTemporaryPinnedApps: false,
  search: {
    query: undefined,
    result: [],
  },
}

export const appStateSlice = createSlice({
  name: 'appState',
  initialState,
  reducers: {
    // GlobalAppState
    setDisplayAllApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displayAllApps = payload
    },
    setDisplaySettings: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySettings = payload
    },
    setMenuAppDetails: (state: AppState, { payload }: PayloadAction<AppDetails | undefined>) => {
      state.menuAppDetails = payload
    },
    setDisplayAppMenu: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displayAppMenu = payload
    },
    setDisplaySortableFavoriteApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortableFavoriteApps = payload
    },
    setDisplaySortablePinnedApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortablePinnedApps = payload
    },
    setDisplaySortableTemporaryPinnedApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortableTemporaryPinnedApps = payload
    },
    // AppsSearchState
    setAppsSearchQuery: (state: AppState, { payload }: PayloadAction<string | undefined>) => {
      state.search.query = payload
    },
    setAppsSearchResult: (state: AppState, { payload }: PayloadAction<AppDetails[]>) => {
      state.search.result = payload
    },
    resetAppsSearchState: (state: AppState) => {
      state.search.query = undefined
      state.search.result = []
    },
  },
})

export const {
  setDisplayAllApps,
  setDisplaySettings,
  setMenuAppDetails,
  setDisplayAppMenu,
  setDisplaySortableFavoriteApps,
  setDisplaySortablePinnedApps,
  setDisplaySortableTemporaryPinnedApps,
  setAppsSearchQuery,
  setAppsSearchResult,
  resetAppsSearchState,
} = appStateSlice.actions

export const toogleAllApps = createAction('appState/toogleAllApps')
export const appLaunch = createAction<{ renderedIn: RenderedIn; appDetails: AppDetails }>('appState/appLaunch')
export const appLaunchFromSearch = createAction('appState/appLaunchFromSearch')
export const sortFavoriteApps = createAction('appState/sortFavoriteApps')
export const sortPinnedApps = createAction('appState/sortPinnedApps')
export const sortTemporaryPinnedApps = createAction('appState/sortTemporaryPinnedApps')

export const selectAppState = (state: RootState) => state.appState
export const selectDisplayAllApps = (state: RootState) => state.appState.displayAllApps
export const selectDisplaySettings = (state: RootState) => state.appState.displaySettings
export const selectMenuAppDetails = (state: RootState) => state.appState.menuAppDetails
export const selectDisplayAppMenu = (state: RootState) => state.appState.displayAppMenu
export const selectDisplaySortableFavoriteApps = (state: RootState) => state.appState.displaySortableFavoriteApps
export const selectDisplaySortablePinnedApps = (state: RootState) => state.appState.displaySortablePinnedApps
export const selectDisplaySortableTemporaryPinnedApps = (state: RootState) => state.appState.displaySortableTemporaryPinnedApps
export const selectAppsSearchQuery = (state: RootState) => state.appState.search.query
export const selectAppsSearchResult = (state: RootState) => state.appState.search.result

export default appStateSlice.reducer
