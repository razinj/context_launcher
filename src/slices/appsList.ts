import { createAction, createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { AppDetails } from '../models/app-details'
import { RootState } from '../store'
import { getAppsLetterIndex } from '../utils/alphabet-list'
import { displayToast } from '../utils/toast'

export interface AppsListState {
  list: AppDetails[]
}

const initialState: AppsListState = {
  list: [],
}

export const appsListSlice = createSlice({
  name: 'appsList',
  initialState,
  reducers: {
    setAppsList: (state: AppsListState, { payload }: PayloadAction<AppDetails[]>) => {
      // console.log('setAppsList called');
      displayToast(`apps found : ${payload.length}`)

      state.list = payload.sort((appOne: AppDetails, appTwo: AppDetails) => appOne.name.localeCompare(appTwo.name))
    },
  },
})

export const { setAppsList } = appsListSlice.actions

export const getAppsListAction = createAction('appsList/getAppsListAction')
export const appRemovedAction = createAction<string>('appsList/appRemovedAction')
export const revalidateAppsLists = createAction('appsList/revalidateAppsLists')

const selectAppsList = (state: RootState) => state.appsList.list

export const selectAppsListMemoized = createSelector(selectAppsList, (list: AppDetails[]) => list)
export const selectAppsLetterListMemoized = createSelector(selectAppsList, (list: AppDetails[]) =>
  getAppsLetterIndex(list)
)

export default appsListSlice.reducer
