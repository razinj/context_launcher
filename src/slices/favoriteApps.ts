// Redux
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
// Models
import { RootState } from '../store'
import { FavoriteApp } from '../models/favorite-app'

export interface FavoriteAppsState {
  list: FavoriteApp[]
}

const initialState: FavoriteAppsState = {
  list: [],
}

export const favoriteAppsSlice = createSlice({
  name: 'favoriteApps',
  initialState,
  reducers: {
    addFavoriteApp: (state: FavoriteAppsState, { payload }: PayloadAction<FavoriteApp>) => {
      if (state.list.length == 5) return

      const existingApp = state.list.find(({ name }: FavoriteApp) => name === payload.name)

      // Add to list only if it doesn't exists
      if (!existingApp) state.list.push({ ...payload })
    },
    removeFavoriteApp: (state: FavoriteAppsState, { payload }: PayloadAction<string>) => {
      const foundAppIndex = state.list.findIndex(({ name }: FavoriteApp) => name === payload)

      if (foundAppIndex === -1) return

      state.list.splice(foundAppIndex, 1)
    },
    setFavoriteApps: (state: FavoriteAppsState, { payload }: PayloadAction<FavoriteApp[]>) => {
      state.list = payload
    },
    clearFavoriteApps: (state: FavoriteAppsState) => {
      state.list = []
    },
  },
})

export const { addFavoriteApp, removeFavoriteApp, setFavoriteApps, clearFavoriteApps } = favoriteAppsSlice.actions

const selectFavoriteApps = (state: RootState) => state.favoriteApps.list

export const selectFavoriteAppsMemoized = createSelector(selectFavoriteApps, (list: FavoriteApp[]) => list)
export const selectFavoriteAppsCountMemoized = createSelector(selectFavoriteApps, (list: FavoriteApp[]) => list.length)

export default favoriteAppsSlice.reducer
