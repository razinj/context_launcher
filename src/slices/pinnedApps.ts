import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { PinnedApp, PinnedAppsState, TemporaryPinnedAppsConfig } from '../models/pinned-app'
import { RootState } from '../store'
import { getAppIndex } from '../utils/apps'

const initialState: PinnedAppsState = {
  list: [],
  temporarily: [], // Temporarily pinned apps
  temporaryAppsConfig: {
    startDate: undefined,
    endDate: undefined,
  },
}

export const pinnedAppsSlice = createSlice({
  name: 'pinnedApps',
  initialState,
  reducers: {
    addPinnedApp: (state: PinnedAppsState, { payload }: PayloadAction<{ app: PinnedApp; isPermanent: boolean }>) => {
      if (payload.isPermanent) {
        const foundAppIndex = getAppIndex(state.list, payload.app.packageName)

        if (foundAppIndex !== -1) return

        state.list.push({ ...payload.app })
      } else {
        const foundAppIndex = getAppIndex(state.temporarily, payload.app.packageName)

        if (foundAppIndex !== -1) return

        state.temporarily.push({ ...payload.app })
      }
    },
    removePinnedApp: (state: PinnedAppsState, { payload }: PayloadAction<{ app: PinnedApp; isPermanent: boolean }>) => {
      if (payload.isPermanent) {
        const foundAppIndex = getAppIndex(state.list, payload.app.packageName)

        if (foundAppIndex === -1) return

        state.list.splice(foundAppIndex, 1)
      } else {
        const foundAppIndex = getAppIndex(state.temporarily, payload.app.packageName)

        if (foundAppIndex === -1) return

        state.temporarily.splice(foundAppIndex, 1)
      }
    },
    clearPinnedApps: (state: PinnedAppsState, { payload }: PayloadAction<{ temporarily: boolean }>) => {
      if (payload.temporarily) state.temporarily = []
      else state.list = []
    },
    setTemporaryAppsConfig: (state: PinnedAppsState, { payload }: PayloadAction<TemporaryPinnedAppsConfig>) => {
      state.temporaryAppsConfig = { ...payload }
    },
    setPinnedApps: (
      state: PinnedAppsState,
      { payload }: PayloadAction<{ apps: PinnedApp[]; isPermanent: boolean }>
    ) => {
      if (payload.isPermanent) state.list = payload.apps
      else state.temporarily = payload.apps
    },
  },
})

export const { addPinnedApp, removePinnedApp, setTemporaryAppsConfig, clearPinnedApps, setPinnedApps } =
  pinnedAppsSlice.actions

const selectPinnedApps = (state: RootState) => state.pinnedApps.list
const selecttTemporaryPinnedApps = (state: RootState) => state.pinnedApps.temporarily
const selectTemporaryPinnedAppsConfig = (state: RootState) => state.pinnedApps.temporaryAppsConfig

export const selectPinnedAppsMemoized = createSelector(selectPinnedApps, (list: PinnedApp[]) => list)
export const selectPinnedAppsCountMemoized = createSelector(selectPinnedApps, (list: PinnedApp[]) => list.length)
export const selectTemporaryPinnedAppsMemoized = createSelector(selecttTemporaryPinnedApps, (list: PinnedApp[]) => list)
export const selectTemporaryPinnedAppsCountMemoized = createSelector(
  selecttTemporaryPinnedApps,
  (list: PinnedApp[]) => list.length
)
export const selectTemporaryPinnedAppsConfigMemoized = createSelector(
  selectTemporaryPinnedAppsConfig,
  (config: TemporaryPinnedAppsConfig) => config
)

export default pinnedAppsSlice.reducer
