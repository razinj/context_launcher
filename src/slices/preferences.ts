// Redux
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
// Models
import { RootState } from '../store'

export interface PreferencesState {
  displayRecentApps: boolean
  displayFavoriteApps: boolean
}

const initialState: PreferencesState = {
  displayRecentApps: true,
  displayFavoriteApps: true,
}

export const preferencesSlice = createSlice({
  name: 'preferences',
  initialState,
  reducers: {
    displayRecentApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayRecentApps = payload
    },
    displayFavoriteApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayFavoriteApps = payload
    },
    resetPreferences: (state: PreferencesState) => {
      state.displayRecentApps = true
      state.displayFavoriteApps = true
    },
  },
})

export const { displayRecentApps, displayFavoriteApps, resetPreferences } = preferencesSlice.actions

const selectDisplayRecentApps = (state: RootState) => state.preferences.displayRecentApps
const selectDisplayFavoriteApps = (state: RootState) => state.preferences.displayFavoriteApps

export const selectDisplayRecentAppsMemoized = createSelector(selectDisplayRecentApps, (value: boolean) => value)
export const selectDisplayFavoriteAppsMemoized = createSelector(selectDisplayFavoriteApps, (value: boolean) => value)

export default preferencesSlice.reducer
