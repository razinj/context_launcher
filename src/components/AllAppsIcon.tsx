import React from 'react'
import { IconButton } from 'react-native-paper'
import { useDispatch } from 'react-redux'
import { PRIMARY_COLOR } from '../constants'
import { iconButtonStyle } from '../shared/bottom-container'
import { toogleAllApps } from '../slices/appState'

const AllAppsIcon = () => {
  const dispatch = useDispatch()

  const _toogleAllApps = () => {
    dispatch(toogleAllApps())
  }

  return (
    <IconButton
      size={30}
      icon='hexagon-outline'
      style={iconButtonStyle}
      iconColor={PRIMARY_COLOR}
      onPress={_toogleAllApps}
      testID='all-apps-toggle-button'
    />
  )
}

export default AllAppsIcon
