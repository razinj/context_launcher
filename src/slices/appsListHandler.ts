import { PayloadAction } from '@reduxjs/toolkit'
import { Channel, channel } from 'redux-saga'
import { call, fork, put, select, take, takeLatest } from 'redux-saga/effects'
import { AppDetails } from '../models/app-details'
import LauncherAppsModule from '../native-modules/LauncherAppsModule'
import { appRemovedAction, getAppsListAction, revalidateAppsLists, setAppsList } from './appsList'
import { removeFavoriteApp, removeFavoriteApps, selectFavoriteAppsMemoized } from './favoriteApps'
import {
  removePinnedApp,
  removePinnedApps,
  selectPinnedAppsMemoized,
  selectTemporaryPinnedAppsMemoized,
} from './pinnedApps'
import { removeRecentApp, removeRecentApps, selectRecentAppsMemoized } from './recentApps'

const notAvailable = 'NOT_AVAILABLE'

function* getAppsListActionHandler() {
  try {
    const applications: string = yield call(LauncherAppsModule.getApplications)
    const apps = JSON.parse(applications) as AppDetails[]
    yield put(setAppsList(apps))
  } catch (error: unknown) {
    yield put(setAppsList([]))
    console.error('Error getting apps list:', error)
  }
}

function* appRemovedActionHandler(action: PayloadAction<string>) {
  yield put(removeRecentApp(action.payload))
  const pinnedApp = { id: notAvailable, packageName: action.payload, name: notAvailable, icon: notAvailable }
  yield put(removePinnedApp({ app: pinnedApp, isPermanent: true })) // TODO: Check why I send an "app" inseatd of "packageName" for removal
  yield put(removePinnedApp({ app: pinnedApp, isPermanent: false }))
  yield put(removeFavoriteApp(action.payload))
}

// Revalidate all apps' lists
function* x() {
  console.log('x called')

  const applications: string = yield call(LauncherAppsModule.getApplications)
  const apps = JSON.parse(applications) as AppDetails[]
  yield put(setAppsList(apps))

  const appsPackageNames = apps.map(({ packageName }: AppDetails) => packageName)

  // Revalidate recent apps
  const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
  yield put(
    removeRecentApps(
      recentApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate favorite apps
  const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
  yield put(
    removeFavoriteApps(
      favoriteApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate pinned apps
  const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: true,
      packageNames: pinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )

  // Revalidate temporary pinned apps
  const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: false,
      packageNames: temporaryPinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )
}

// const upsertAppsListChannel = channel()

// function* watchAppsListUpsert() {
//   while (true) {
//     type x = typeof setAppsList
//     const action: x = yield take(upsertAppsListChannel)
//     yield put(action)
//   }
// }

function getApplicationsV2Async() {
  return new Promise((resolve, reject) => {
    LauncherAppsModule.getApplicationsV2((applications: string) => {
      resolve(applications)
    })
  })
}

function* y() {
  console.log('y called')

  const applications: string = yield call(getApplicationsV2Async)
  yield call(handleApplications, applications)

  const apps = JSON.parse(applications) as AppDetails[]

  const appsPackageNames = apps.map(({ packageName }: AppDetails) => packageName)
  // const appsPackageNames: string[] = []

  // Revalidate recent apps
  const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
  yield put(
    removeRecentApps(
      recentApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate favorite apps
  const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
  yield put(
    removeFavoriteApps(
      favoriteApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate pinned apps
  const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: true,
      packageNames: pinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )

  // Revalidate temporary pinned apps
  const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: false,
      packageNames: temporaryPinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )
}

function* handleApplications(applications: string) {
  console.log('handleApplications called')

  const apps = JSON.parse(applications) as AppDetails[]
  // dispatch an action here
  yield put(setAppsList(apps))
}

function* childSaga(channel: Channel<any>) {
  const applications: string = yield call(LauncherAppsModule.getApplications)
  const apps = JSON.parse(applications) as AppDetails[]
  // yield put(setAppsList(apps))
  // const result = yield call(someAsyncFunction);
  yield put(channel, apps) // Emit the result through the channel
}

function* z() {
  console.log('z called')

  const communicationChannel: Channel<AppDetails[]> = yield call(channel)
  yield fork(childSaga, communicationChannel)

  while (true) {
    const result: AppDetails[] = yield take(communicationChannel)
    console.log(' result: ', result.length)

    const appsPackageNames = result.map(({ packageName }: AppDetails) => packageName)
    // const appsPackageNames: string[] = []

    // Revalidate recent apps
    const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
    yield put(
      removeRecentApps(
        recentApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          )
      )
    )

    // Revalidate favorite apps
    const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
    yield put(
      removeFavoriteApps(
        favoriteApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          )
      )
    )

    // Revalidate pinned apps
    const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
    yield put(
      removePinnedApps({
        isPermanent: true,
        packageNames: pinnedApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          ),
      })
    )

    // Revalidate temporary pinned apps
    const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
    yield put(
      removePinnedApps({
        isPermanent: false,
        packageNames: temporaryPinnedApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          ),
      })
    )

    // Use the result returned by the child saga
    // Non-blocking logic here

    // ...
  }

  console.log('done')
}

export function* appsListSagas() {
  yield takeLatest(getAppsListAction.type, getAppsListActionHandler)
  yield takeLatest(appRemovedAction.type, appRemovedActionHandler)
  // yield takeLatest(revalidateAppsLists.type, x)
  // yield takeLatest(revalidateAppsLists.type, z)
  yield takeLatest(revalidateAppsLists.type, z)
}
