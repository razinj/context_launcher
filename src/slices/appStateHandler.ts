import { PayloadAction } from '@reduxjs/toolkit'
import { call, put, select, takeLatest } from 'redux-saga/effects'
import { AppDetails } from '../models/app-details'
import { FavoriteApp } from '../models/favorite-app'
import { PinnedApp } from '../models/pinned-app'
import { RenderedIn } from '../models/rendered-in'
import { launchApp } from '../utils/apps-module'
import { dismissKeyboard } from '../utils/keyboard'
import {
  appLaunch,
  appLaunchFromSearch,
  resetAppsSearchState,
  selectAppsSearchResult,
  selectDisplayAllApps,
  setDisplayAllApps,
  setDisplaySettings,
  setDisplaySortableFavoriteApps,
  setDisplaySortablePinnedApps,
  setDisplaySortableTemporaryPinnedApps,
  setMenuAppDetails,
  sortFavoriteApps,
  sortPinnedApps,
  sortTemporaryPinnedApps,
  toogleAllApps,
} from './appState'
import { selectFavoriteApps } from './favoriteApps'
import { selectPinnedApps } from './pinnedApps'
import { addRecentApp } from './recentApps'

// We should only add a launched app if it was clicked (rendered) in one of the following places;
const ADD_TO_RECENT_APPS_RENDERED_IN_VALUES = [RenderedIn.ALL_APPS, RenderedIn.FILTERED_APPS]

function* toggleAllAppsHandler() {
  const allAppsDisplayed: boolean = yield select(selectDisplayAllApps)
  yield put(setDisplayAllApps(!allAppsDisplayed))
}

function* appLaunchHandler(action: PayloadAction<{ renderedIn: RenderedIn; appDetails: AppDetails }>) {
  yield call(dismissKeyboard)
  yield put(resetAppsSearchState())
  yield put(setDisplayAllApps(false))
  yield put(setMenuAppDetails(undefined))
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortableFavoriteApps(false))

  const { appDetails, renderedIn } = action.payload

  if (!ADD_TO_RECENT_APPS_RENDERED_IN_VALUES.includes(renderedIn)) {
    return
  }

  // Skip adding to recent apps if permanently pinned/favourited but launched from different lists/views
  const pinnedApps: PinnedApp[] = yield select(selectPinnedApps)
  const favoriteApps: FavoriteApp[] = yield select(selectFavoriteApps)
  // TODO: Include the temporarily pinned apps but only when they're rendered/displayed.

  const pinnedAppsPackages: string[] = pinnedApps.map(({ packageName }: PinnedApp) => packageName)
  const favoriteAppsPackages: string[] = favoriteApps.map(({ packageName }: FavoriteApp) => packageName)
  const allApps = new Set([...pinnedAppsPackages, ...favoriteAppsPackages])

  if (allApps.has(appDetails.packageName)) {
    return
  }

  // Otherwise add to recent apps
  yield put(addRecentApp(appDetails))
}

function* sortFavoriteAppsHandler() {
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortableFavoriteApps(true))
}

function* sortPinnedAppsHandler() {
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortablePinnedApps(true))
}

function* sortTemporaryPinnedAppsHandler() {
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortableTemporaryPinnedApps(true))
}

function* appLaunchFromSearchHandler() {
  const searchResult: AppDetails[] = yield select(selectAppsSearchResult)
  if (searchResult.length === 0) {
    return
  }

  const appDetails = searchResult[0]

  yield call(launchApp, appDetails.packageName)
  yield put(appLaunch({ renderedIn: RenderedIn.FILTERED_APPS, appDetails }))
}

export function* appStateSaga() {
  yield takeLatest(toogleAllApps.type, toggleAllAppsHandler)
  yield takeLatest(appLaunch.type, appLaunchHandler)
  yield takeLatest(sortFavoriteApps.type, sortFavoriteAppsHandler)
  yield takeLatest(sortPinnedApps.type, sortPinnedAppsHandler)
  yield takeLatest(sortTemporaryPinnedApps.type, sortTemporaryPinnedAppsHandler)
  yield takeLatest(appLaunchFromSearch.type, appLaunchFromSearchHandler)
}
