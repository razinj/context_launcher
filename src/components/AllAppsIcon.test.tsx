import { fireEvent, screen } from '@testing-library/react-native'
import React from 'react'
import { renderWithProvider, renderWithProviderAndContexts } from '../../utils/test/utils'
import { toogleAllApps } from '../slices/appState'
import AllAppsIcon from './AllAppsIcon'

const useDispatchMock = jest.fn()

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => useDispatchMock,
}))

describe('<AllAppsIcon /> Tests', () => {
  it('should render correctly and match snapshot', () => {
    renderWithProvider(<AllAppsIcon />)

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('all-apps-toggle-button')).toBeOnTheScreen()
  })

  it('should call function to toggle all apps display when pressed', () => {
    renderWithProviderAndContexts(<AllAppsIcon />, {})

    const allAppsButton = screen.getByTestId('all-apps-toggle-button')

    expect(allAppsButton).toBeOnTheScreen()

    fireEvent.press(allAppsButton)

    expect(useDispatchMock).toBeCalledWith(toogleAllApps())
  })
})
