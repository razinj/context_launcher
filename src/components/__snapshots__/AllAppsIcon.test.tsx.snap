// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AllAppsIcon /> Tests should render correctly and match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="all-apps-toggle-button"
  >
    <Icon
      color="#42855B"
      name="hexagon-outline"
      size={34}
      style={
        {
          "margin": 5,
        }
      }
    />
  </View>
</View>
`;
